# Snakemake workflow for ONT WGS and Adaptive Sampling Alignment

This workflow uses unaligned called bam files of ONT data to produce aligned bam files, SNV calls, SV calls, CNV calls, VEP annotations, and QC statistics. All input bams should be checked for quality control prior to using this workflow.

The master branch is set up to use conda environments assumed to be on the user's path. Switch to the `conda_explicit` branch to use the .yamls in the /env directory and then modify the `config.yaml` on that branch. You could also use the yamls to make conda environments on your system and then use the `config_mcclintock.yaml` on the master branch. You should only do this if running on a new server. 

## Usage
```
snakemake -p --use-conda --cores 30
```

### Setup

1. Clone this repository to your **final destination for files**. Snakemake works best with relative paths.
2. Edit `Snakefile` to include the appropriate rules. Most likely this is already set up correctly. `get_targets` generates alignment files for all samples in the file `config/targets.txt`.
    - The workflow searches for bam files with a matching prefix (M[0-9]{5}) in the directory (`input_dir`) specified in the config file.
    - Optionally change `explicitLibraries` to `true` to specify a list of bam files in the `config/targets.txt` instead
    - choose to include either `config/config_mcclintock.yaml` or `config/config_franklin.yaml` depending on which server you would like to run on.
3. Create a sample database file, like  `config/NSC0000_example_config.tsv` with your sample information. All fields are required. 
4. Edit `config/targets.txt` to include only the samples you wish to generate alignments for.
5. Edit `config/config_mcclintock.yaml` or `config/config_franklin.yaml` depending on which server you would like to run on.
    - if running on a new server, update paths for all reference data, tools, and input/output directories.
6. Run!

### Useful flags and recommendations

If work is interrupted or resumed after failure, use `--rerun-triggers mtime` to not re-generate already completed files and/or `--rerun-incomplete` to remove incomplete/corrupted files.
If mamba is not installed, use `--conda-frontend conda`.
For best performance, devote set `threads` to 30 in the config file and run with at least 1 additional core per group of 30 threads, e.g. if `threads: 30`, then run `snakemake --cores 62`. This allows non-processor intensive (but time intensive!) tasks to run in the background of tasks that are both processor and time intensive.

## Outputs

- Indexed bam file of full alignmentIndexed bam file of KIF1A only, plus sex check controls FMR1 and COL1A1
- SNV vcf generated with clair3, phased and not phased
- SV vcfs, phased and not phased
    - generated by sniffles
    - generated by SVIM
    - generated by cuteSV
- CNV vcf and plots, generated by QDNAseq
- VEP vcf of SNV annotations
    - csv containing variants with population allele frequencies < 0.01
- quality control statistics
    - cramino stats for full alignment includes data yield, N50, etc.
        - cramino stats for target region includes more accurate estimate of read lengths for target region only
    - phasing stats, generated with whatshap
    - hp_dp stats, the average read depth per haplotagÂ in targeted regions.

## Requirements

Snakemake 7.32.4 (compatible with Snakemake versions > 8)
Python >= 3.10
